libnavigation-base/src/main/java/com/mapbox/navigation/base/route/model/RouteProgressNavigation.kt

```diff
@@ -0,0 +1,323 @@
+     *
+     * @return list of points representing the current step
+     */
+    fun currentStepPoints() = currentStepPoints
+
+    /**
+     * Provides a list of points that represent the upcoming step
+     * step geometry.
+     *
+     * @return list of points representing the upcoming step
+     */
+    fun upcomingStepPoints() = upcomingStepPoints
+
+    /**
+     * Returns whether or not the location updates are
+     * considered in a tunnel along the route.
+     *
+     * @return true if in a tunnel, false otherwise
+     */
+    fun inTunnel() = inTunnel
+
+    /**
+     * Returns the current state of progress along the route.  Provides route and location tracking
+     * information.
+     *
+     * @return the current state of progress along the route.
+     */
+    fun currentState() = currentState
+
+    /**
+     * Returns the current [DirectionsRoute] geometry with a buffer
+     * that encompasses visible tile surface are while navigating.
+     *
+     *
+     * This [Geometry] is ideal for offline downloads of map or routing tile
+     * data.
+     *
+     * @return current route geometry with buffer
+     */
+    fun routeGeometryWithBuffer() = routeGeometryWithBuffer
+
+    fun toBuilder() = builder
+
+    internal fun currentStep() = currentStep
+
+    internal fun stepIndex() = stepIndex
+
+    internal fun legDistanceRemaining() = legDistanceRemaining
+
+    internal fun stepDistanceRemaining() = stepDistanceRemaining
+
+    internal fun legDurationRemaining() = legDurationRemaining
+
+    class Builder {
+        private var directionsRoute: Route? = null
+        private var legIndex: Int? = null
+        private var distanceRemaining: Double? = null
+        private var currentLegProgress: RouteLegProgressNavigation? = null
+        private var currentStepPoints: List<Point>? = null
+        private var upcomingStepPoints: List<Point>? = null
+        private var inTunnel: Boolean? = null
+        private var currentState: RouteProgressStateNavigation? = null
+        private var routeGeometryWithBuffer: Geometry? = null
+        private var currentStep: LegStepNavigation? = null
+        private var stepIndex: Int? = null
+        private var legDistanceRemaining: Double? = null
+        private var stepDistanceRemaining: Double? = null
+        private var legDurationRemaining: Double? = null
+
+        fun directionsRoute(directionsRoute: Route) =
+                apply { this.directionsRoute = directionsRoute }
+
+        fun legIndex(legIndex: Int) = apply { this.legIndex = legIndex }
+
+        fun distanceRemaining(distanceRemaining: Double) =
+                apply { this.distanceRemaining = distanceRemaining }
+
+        fun currentLegProgress(currentLegProgress: RouteLegProgressNavigation) =
+                apply { this.currentLegProgress = currentLegProgress }
+
+        fun currentStepPoints(currentStepPoints: List<Point>?) =
+                apply { this.currentStepPoints = currentStepPoints }
```

Comment:
Why do we set **step** data here and duplicate it when creating legs? Wouldn't it be more intuitive if each step held its own points and the user could call `RouteProgressNavigation.currentLeg.currentStep.points`?

Reply:
I am not completely aware of how the JSON response received from the Server API is structured and how does one or the other way of setting `step` affects it. I will let @Guardiola31337 answer this

Reply:
It doesn't affect it in any way as `RouteProgress` is an object created and populated by the SDK 👀 https://github.com/mapbox/mapbox-navigation-android/blob/f5c36c93662fafed3e0a57afdc0bfd5445f2edc2/libandroid-navigation/src/main/java/com/mapbox/services/android/navigation/v5/internal/navigation/NavigationRouteProcessor.kt#L89-L107

I believe all this comes from keeping parity with the legacy [`RouteProgress`](https://github.com/mapbox/mapbox-navigation-android/blob/master/libandroid-navigation/src/main/java/com/mapbox/services/android/navigation/v5/routeprogress/RouteProgress.kt) (correct me @abhishek1508 if I'm wrong) and maybe was implemented this way in the first place for convenience and ease of usage and avoid long [Message Chains](https://refactoring.guru/smells/message-chains)? Don't know as `RouteProgress` has been implemented since forever...

BTW, we should definitely take advantage of the refactor and structure the data differently and provide better APIs.

Reply:
Yes this was implemented to keep the parity with `RouteProgress`